src/hotspot/share/ci/ciField.cpp

```diff
@@ -218,28 +223,57 @@ ciField::ciField(fieldDescriptor *fd) :
 
 // Special copy constructor used to flatten inline type fields by
 // copying the fields of the inline type to a new holder klass.
-ciField::ciField(ciField* field, ciInstanceKlass* holder, int offset, bool is_final) {
-  assert(field->holder()->is_inlinetype() || field->holder()->is_abstract(), "should only be used for inline type field flattening");
-  // Set the is_final flag
-  jint final = is_final ? JVM_ACC_FINAL : ~JVM_ACC_FINAL;
-  AccessFlags flags(field->flags().as_int() & final);
-  _flags = ciFlags(flags);
-  _holder = holder;
-  _offset = offset;
-  // Copy remaining fields
-  _name = field->_name;
-  _signature = field->_signature;
-  _type = field->_type;
-  // Trust final flat fields
-  _is_constant = is_final;
-  _known_to_link_with_put = field->_known_to_link_with_put;
-  _known_to_link_with_get = field->_known_to_link_with_get;
-  _constant_value = field->_constant_value;
-  assert(!field->is_flat(), "field must not be flat");
+ciField::ciField(ciField* declared_field, ciField* subfield) {
+  assert(subfield->holder()->is_inlinetype() || subfield->holder()->is_abstract(), "should only be used for inline type field flattening");
+  assert(!subfield->is_flat(), "subfield must not be flat");
+  assert(declared_field->is_flat(), "declared field must be flat");
+
+  _flags = declared_field->flags();
+  _holder = declared_field->holder();
+  _offset = declared_field->offset_in_bytes() + (subfield->offset_in_bytes() - declared_field->type()->as_inline_klass()->payload_offset());
+
+  char buffer[256];
+  jio_snprintf(buffer, sizeof(buffer), "%s.%s", declared_field->name()->as_utf8(), subfield->name()->as_utf8());
```

Comment:
There's some code that matches on `ciField::name`, for example in `ciField::is_call_site_target` but I think that should never be used for flat fields.